File: src/main/java/aoc2021/day04/Bingo.java
package aoc2021.day04;

import library.Year;
import library.io.Input;
import library.tuple.Pair;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Bingo {

    private static Pair<List<Board>, List<Integer>> read() {
        Scanner sc = Input.openFile(Year.AOC_2021, "day04.txt");

        List<Integer> draws = Arrays.stream(sc.nextLine().split(","))
                .map(Integer::parseInt)
                .collect(Collectors.toList());

        List<Board> boards = new ArrayList<>();
        while (sc.hasNext()) {
            sc.nextLine();
            boards.add(Board.read(sc, 5));
        }

        return new Pair<>(boards, draws);
    }

    private static Pair<Board, Integer> winner(List<Board> boards, List<Integer> draws) {
        for (Integer num : draws) {
            for (Board board : boards) {
                board.setNumbers(num);
                if (board.won()) {
                    return new Pair<>(board, num);
                }
            }
        }
        return null;
    }

    private static Pair<Board, Integer> loser(List<Board> boards, List<Integer> draws) {
        Pair<Board, Integer> winner = winner(boards, draws);

        boards.remove(winner.getFirst());

        if (boards.size() == 1) {
            return winner(boards, draws);
        }
        return loser(boards, draws);
    }

    private static void printPair(Pair<Board, Integer> pair) {
        System.out.println(pair.getFirst().getScore() * pair.getSecond());
    }

    public static void main(String[] args) {
        // load data
        Pair<List<Board>, List<Integer>> data = read();
        List<Board> boards = data.getFirst();
        List<Integer> draws = data.getSecond();

        // part 1
        printPair(winner(boards, draws));

        // part 2
        printPair(loser(boards, draws));
    }

}

File: kotlin/src/main/kotlin/aoc2021/day04/Bingo.kt
package aoc2021.day04

import library.Year
import library.head
import library.readData
import library.tail

class Board(values: List<List<Int>>) {
    val marks: List<List<Boolean>> = List(values.size) { List(values[0].size) { false } }
}

fun parseBoard(lines: List<String>): Pair<Board, List<String>> =
     Pair(
        Board(lines.takeWhile { it != "" }
            .map { line -> line.split(" ").map { it.toInt() } }),
        lines.dropWhile { it != "" }.tail()
    )

fun parseBoards(lines: List<String>): List<Board> {
    when (lines.size) {
        0 -> return emptyList()
        else -> {
            val (board, next) = parseBoard(lines)
            return listOf(board) + parseBoards(next)
        }
    }
}

fun parse(lines: List<String>): Pair<List<Int>, List<Board>> =
    Pair(
        lines.head().split(",").map { it.toInt() },
        parseBoards(lines.tail().tail())
    )

fun main() {
    val data = readData(Year._2021, "day04_e.txt")

    parse(data)
}